prog/vm/nexus.rb

```diff
@@ -29,18 +29,44 @@ def self.assemble(public_key, project_id, name: nil, size: "m5a.2x",
     key_wrapping_key = cipher.random_key
     key_wrapping_iv = cipher.random_iv
 
-    # if the caller hasn't provided any subnets, generate a random one
-    if private_subnets.empty?
-      private_subnets.append([random_ula, random_private_ipv4])
-    end
-
     DB.transaction do
+      # Here the logic is the following;
+      # - If the user provided nic_id, that nic has to exist and we fetch private_subnet
+      # from the reference of nic. We just assume it and not even check the validity of the
+      # private_subnet_id.
+      # - If the user did not provide nic_id but the private_subnet_id, that private_subnet
+      # must exist, otherwise we fail.
+      # - If the user did not provide nic_id but the private_subnet_id and that subnet exists
+      # then we create a nic on that subnet.
+      # - If the user provided neither nic_id nor private_subnet_id, that's OK, we create both.
+      nic = nil
+      subnet = nil
+      if nic_id
+        nic = Nic[nic_id]
```

Comment:
What happens if user passes id of another customer?

Reply:
That's a good question, though, right now, we don't have much mechanism around entity ownership tied to a customer (organisation?). We have projects and subnets are tied to a project, so I think I can use it. The https://github.com/ubicloud/ubicloud/pull/334 actually implements something similar, I will get that pr in shortly after this one

Reply:
well, we don't have entity ownership as such because we don't have an owner role per se, because we're ABAC. This may be an @enescakir kind of question, though it should prompt survey of comparison implementations (the hyperscalers).

We may not do such authorizations in nexuses anyway, it may live in the API, though I sort of think defense-in-depth/avoiding horrible operator mistakes even at the cost of re-checking authorization is a good idea.

Reply:
Private subnets are associated with projects. So we can find nic in nics of private subnets of project instead of all nics.

Previous resources (vm, project) do authorization checks at routes/controller level. It blocks customers to access other customers resources, but not operator.

If we want to avoid operator mistakes and move some kind of authorization check in progs, we need to think more about it. Because we need a subject for authorization checks. It's customer at routes/controllers. What will be subject in progs? Because operator doesn't have permissions to access objects. So our current authorization code doesn't work without a subject.

Reply:
@enescakir can you check the latest version of the operation here.

Reply:
It seems avoiding operator mistakes requires significant investment right now. I think similar type of investment might be needed at the time of building an admin panel. I'd suggest leaving that to there. If we can check the authorization in the API level to prevent potential vulnerability, it is a win.

Reply:
Yeah. I think it's better now. We need to add extra authorization check in you UI PR.

Reply:
Thanks